File: android/src/main/java/it/hubagency/spatialscan/RoomHistoryManager.kt
package it.hubagency.spatialscan

import android.content.Context
import android.util.Log
import com.getcapacitor.JSObject
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.util.UUID

/**
 * Persistenza locale dello storico stanze.
 *
 * Formato: filesDir/hub_rooms.json — array JSON di RoomRecord.
 * Nessuna dipendenza esterna: usa org.json (incluso in Android SDK).
 */
object RoomHistoryManager {

    private const val TAG      = "RoomHistoryManager"
    private const val FILENAME = "hub_rooms.json"

    // ── API pubblica ──────────────────────────────────────────────────────────

    /**
     * Salva una nuova stanza estraendo i metadati dal JSObject restituito da buildResult().
     * Se il salvataggio fallisce, logga l'errore senza propagare eccezioni.
     */
    fun save(context: Context, result: JSObject, name: String): RoomRecord? {
        return try {
            val record = buildRecord(result, name)
            val list   = loadRaw(context).toMutableList()
            list.add(0, record)          // più recente in cima
            writeAll(context, list)
            saveRoomData(context, record.id, result)
            Log.d(TAG, "saved room '${record.name}' id=${record.id} area=${record.area}m²")
            record
        } catch (e: Exception) {
            Log.e(TAG, "save failed: ${e.message}", e)
            null
        }
    }

    /** Restituisce tutti i record in ordine dal più recente. */
    fun loadAll(context: Context): List<RoomRecord> {
        return try {
            loadRaw(context)
        } catch (e: Exception) {
            Log.e(TAG, "loadAll failed: ${e.message}", e)
            emptyList()
        }
    }

    /**
     * Aggiorna bilateralmente il metadata di un'apertura in una stanza già salvata.
     * Usato quando la stanza adiacente collega una porta precedentemente PENDING.
     * Aggiorna sia hub_rooms.json che hub_room_{id}.json.
     */
    fun updateOpeningMetadata(
        context:        Context,
        roomId:         String,
        openingId:      String,
        linkedRoomId:   String,
        linkedRoomName: String
    ) {
        try {
            // 1. Aggiorna hub_room_{id}.json (dati geometrici completi)
            val roomData = loadRoomData(context, roomId)
            if (roomData != null) {
                val walls = roomData.optJSONArray("walls")
                if (walls != null) {
                    for (i in 0 until walls.length()) {
                        val wall = walls.optJSONObject(i) ?: continue
                        val ops  = wall.optJSONArray("openings") ?: continue
                        for (j in 0 until ops.length()) {
                            val op = ops.optJSONObject(j) ?: continue
                            if (op.optString("id") == openingId) {
                                op.put("isInternal",       true)
                                op.put("connectionStatus", ConnectionStatus.LINKED.name)
                                op.put("linkedRoomId",     linkedRoomId)
                                op.put("connectionLabel",  linkedRoomName)
                            }
                        }
                    }
                }
                File(context.filesDir, "hub_room_$roomId.json").writeText(roomData.toString())
            }

            // 2. Aggiorna hub_rooms.json (indice stanze)
            val list = loadRaw(context).toMutableList()
            val idx  = list.indexOfFirst { it.id == roomId }
            if (idx >= 0) {
                val record      = list[idx]
                val newOpenings = record.openings.toMutableList()
                val existingIdx = newOpenings.indexOfFirst { it.openingId == openingId }
                val wallId      = newOpenings.getOrNull(existingIdx)?.wallId ?: ""
                val newMeta = OpeningMetadata(
                    openingId        = openingId,
                    wallId           = wallId,
                    isInternal       = true,
                    linkedRoomId     = linkedRoomId,
                    connectionLabel  = linkedRoomName,
                    connectionStatus = ConnectionStatus.LINKED
                )
                if (existingIdx >= 0) newOpenings[existingIdx] = newMeta
                else newOpenings.add(newMeta)
                list[idx] = record.copy(openings = newOpenings)
                writeAll(context, list)
            }
            Log.d(TAG, "updateOpeningMetadata roomId=$roomId openingId=$openingId linkedTo=$linkedRoomId")
        } catch (e: Exception) {
            Log.e(TAG, "updateOpeningMetadata failed: ${e.message}", e)
        }
    }

    /** Rimuove un record per id e cancella il file geometrico associato. No-op se non trovato. */
    fun delete(context: Context, id: String) {
        try {
            val list = loadRaw(context).filter { it.id != id }
            writeAll(context, list)
            File(context.filesDir, "hub_room_$id.json").delete()
            Log.d(TAG, "deleted room id=$id")
        } catch (e: Exception) {
            Log.e(TAG, "delete failed: ${e.message}", e)
        }
    }

    // ── Internals ─────────────────────────────────────────────────────────────

    private fun buildRecord(result: JSObject, name: String): RoomRecord {
        val dims     = result.getJSObject("roomDimensions")
        val walls    = result.getJSONArray("walls")
        val floor    = result.getJSObject("floor")

        // Conta aperture e raccoglie metadati collegamento
        var openingCount = 0
        val openingMetas = mutableListOf<OpeningMetadata>()
        if (walls != null) {
            for (i in 0 until walls.length()) {
                val wall   = walls.optJSONObject(i) ?: continue
                val wallId = wall.optString("id")
                val ops    = wall.optJSONArray("openings") ?: continue
                openingCount += ops.length()
                for (j in 0 until ops.length()) {
                    val op = ops.optJSONObject(j) ?: continue
                    if (op.has("isInternal")) {
                        openingMetas.add(OpeningMetadata(
                            openingId       = op.optString("id"),
                            wallId          = wallId,
                            isInternal      = op.optBoolean("isInternal", false),
                            linkedRoomId    = op.optString("linkedRoomId").takeIf { it.isNotEmpty() },
                            connectionLabel = op.optString("connectionLabel").takeIf { it.isNotEmpty() }
                        ))
                    }
                }
            }
        }

        return RoomRecord(
            id            = UUID.randomUUID().toString(),
            name          = name.trim().ifEmpty { "Stanza" },
            timestamp     = System.currentTimeMillis(),
            area          = dims?.getDouble("area")   ?: floor?.getDouble("area") ?: 0.0,
            height        = dims?.getDouble("height") ?: 0.0,
            wallCount     = walls?.length()           ?: 0,
            openingCount  = openingCount,
            floorPlanPath = result.getString("floorPlanPath"),
            glbPath       = result.getString("glbPath"),
            openings      = openingMetas
        )
    }

    private fun file(context: Context): File =
        File(context.filesDir, FILENAME)

    private fun loadRaw(context: Context): List<RoomRecord> {
        val f = file(context)
        if (!f.exists()) return emptyList()
        val json = f.readText()
        val arr  = JSONArray(json)
        return (0 until arr.length()).map { RoomRecord.fromJson(arr.getJSONObject(it)) }
    }

    private fun writeAll(context: Context, records: List<RoomRecord>) {
        val arr = JSONArray()
        records.forEach { arr.put(it.toJson()) }
        file(context).writeText(arr.toString())
    }

    private fun saveRoomData(context: Context, id: String, result: JSObject) {
        try {
            File(context.filesDir, "hub_room_$id.json").writeText(result.toString())
        } catch (e: Exception) {
            Log.e(TAG, "saveRoomData failed id=$id: ${e.message}", e)
        }
    }

    fun loadRoomData(context: Context, id: String): JSONObject? {
        return try {
            val f = File(context.filesDir, "hub_room_$id.json")
            if (!f.exists()) return null
            JSONObject(f.readText())
        } catch (e: Exception) {
            Log.e(TAG, "loadRoomData failed id=$id: ${e.message}", e)
            null
        }
    }
}

File: ios/Plugin/RoomHistoryManager.swift
import Foundation

/// Persistenza locale dello storico stanze.
/// Porta iOS di RoomHistoryManager.kt.
///
/// Formato hub_room_{id}.json:
///   { "success", "roomId", "name", "walls": [{id, startPoint:{x,z}, endPoint:{x,z}, height,
///     openings:[{id, kind, offsetAlongWall, width, height, bottom, isInternal,
///               connectionStatus, linkedRoomId, connectionLabel}]}],
///     "floor": {"vertices":[{x,z}], "area"},
///     "roomDimensions": {width, length, height, area, perimeter} }
///
/// Formato hub_rooms.json: [RoomRecord] (Codable array)
final class RoomHistoryManager {

    static let shared = RoomHistoryManager()
    private init() {}

    private let filename = "hub_rooms.json"

    private var docsDir: URL? {
        FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first
    }

    // MARK: - Save

    /// Salva la stanza: hub_room_{id}.json + aggiorna hub_rooms.json.
    /// roomData: dizionario con chiavi walls, floor, roomDimensions (formato Composer).
    func save(roomData: [String: Any], name: String) -> RoomRecord? {
        guard let docsDir = docsDir else { return nil }
        let id = UUID().uuidString

        let walls = roomData["walls"] as? [[String: Any]] ?? []
        let floor  = roomData["floor"] as? [String: Any]
        let dims   = roomData["roomDimensions"] as? [String: Any]

        var openingCount = 0
        var openingMetas = [OpeningMetadata]()
        for wall in walls {
            let wallId = wall["id"] as? String ?? ""
            let ops    = wall["openings"] as? [[String: Any]] ?? []
            openingCount += ops.count
            for op in ops {
                let statusStr  = op["connectionStatus"] as? String ?? "EXTERNAL"
                let status     = ConnectionStatus(rawValue: statusStr) ?? .external
                openingMetas.append(OpeningMetadata(
                    openingId:        op["id"] as? String ?? "",
                    wallId:           wallId,
                    isInternal:       op["isInternal"] as? Bool ?? false,
                    linkedRoomId:     (op["linkedRoomId"] as? String).flatMap { $0.isEmpty ? nil : $0 },
                    connectionLabel:  (op["connectionLabel"] as? String).flatMap { $0.isEmpty ? nil : $0 },
                    connectionStatus: status
                ))
            }
        }

        let area     = (dims?["area"]   as? Double) ?? (floor?["area"] as? Double) ?? 0.0
        let height   = (dims?["height"] as? Double) ?? 0.0
        let wallCnt  = walls.count

        let record = RoomRecord(
            id:            id,
            name:          name.trimmingCharacters(in: .whitespaces).isEmpty ? "Stanza" : name,
            timestamp:     Int64(Date().timeIntervalSince1970 * 1000),
            area:          area,
            height:        height,
            wallCount:     wallCnt,
            openingCount:  openingCount,
            floorPlanPath: roomData["floorPlanPath"] as? String,
            glbPath:       roomData["glbPath"] as? String,
            openings:      openingMetas
        )

        // hub_room_{id}.json
        var full = roomData
        full["roomId"] = id
        full["success"] = true
        if let data = try? JSONSerialization.data(withJSONObject: full) {
            try? data.write(to: docsDir.appendingPathComponent("hub_room_\(id).json"))
        }

        // hub_rooms.json (più recente in cima)
        var rooms = loadAll()
        rooms.insert(record, at: 0)
        if let data = try? JSONEncoder().encode(rooms) {
            try? data.write(to: docsDir.appendingPathComponent(filename))
        }

        return record
    }

    // MARK: - Load

    func loadAll() -> [RoomRecord] {
        guard let docsDir = docsDir,
              let data    = try? Data(contentsOf: docsDir.appendingPathComponent(filename)),
              let records = try? JSONDecoder().decode([RoomRecord].self, from: data)
        else { return [] }
        return records
    }

    func loadRoomData(id: String) -> [String: Any]? {
        guard let docsDir = docsDir,
              let data    = try? Data(contentsOf: docsDir.appendingPathComponent("hub_room_\(id).json")),
              let obj     = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }
        return obj
    }

    // MARK: - Update opening metadata (bilateral link)

    func updateOpeningMetadata(roomId: String, openingId: String,
                               linkedRoomId: String, linkedRoomName: String) {
        guard let docsDir = docsDir else { return }

        // 1. Aggiorna hub_room_{id}.json
        if var roomData = loadRoomData(id: roomId),
           var walls = roomData["walls"] as? [[String: Any]] {
            for wi in 0..<walls.count {
                if var ops = walls[wi]["openings"] as? [[String: Any]] {
                    for oi in 0..<ops.count {
                        if ops[oi]["id"] as? String == openingId {
                            ops[oi]["isInternal"]       = true
                            ops[oi]["connectionStatus"] = ConnectionStatus.linked.rawValue
                            ops[oi]["linkedRoomId"]     = linkedRoomId
                            ops[oi]["connectionLabel"]  = linkedRoomName
                        }
                    }
                    walls[wi]["openings"] = ops
                }
            }
            roomData["walls"] = walls
            if let data = try? JSONSerialization.data(withJSONObject: roomData) {
                try? data.write(to: docsDir.appendingPathComponent("hub_room_\(roomId).json"))
            }
        }

        // 2. Aggiorna hub_rooms.json
        var rooms = loadAll()
        if let idx = rooms.firstIndex(where: { $0.id == roomId }) {
            var openings = rooms[idx].openings
            let wallId: String
            if let oidx = openings.firstIndex(where: { $0.openingId == openingId }) {
                wallId = openings[oidx].wallId
                openings[oidx] = OpeningMetadata(
                    openingId: openingId, wallId: wallId,
                    isInternal: true, linkedRoomId: linkedRoomId,
                    connectionLabel: linkedRoomName, connectionStatus: .linked)
            } else {
                wallId = ""
                openings.append(OpeningMetadata(
                    openingId: openingId, wallId: wallId,
                    isInternal: true, linkedRoomId: linkedRoomId,
                    connectionLabel: linkedRoomName, connectionStatus: .linked))
            }
            let r = rooms[idx]
            rooms[idx] = RoomRecord(id: r.id, name: r.name, timestamp: r.timestamp,
                                    area: r.area, height: r.height,
                                    wallCount: r.wallCount, openingCount: r.openingCount,
                                    floorPlanPath: r.floorPlanPath, glbPath: r.glbPath,
                                    openings: openings)
            if let data = try? JSONEncoder().encode(rooms) {
                try? data.write(to: docsDir.appendingPathComponent(filename))
            }
        }
    }

    // MARK: - Delete

    func delete(id: String) {
        guard let docsDir = docsDir else { return }
        let rooms = loadAll().filter { $0.id != id }
        if let data = try? JSONEncoder().encode(rooms) {
            try? data.write(to: docsDir.appendingPathComponent(filename))
        }
        try? FileManager.default.removeItem(at: docsDir.appendingPathComponent("hub_room_\(id).json"))
    }

    // MARK: - Clear session

    func clearAll() {
        guard let docsDir = docsDir else { return }
        let fm = FileManager.default
        if let items = try? fm.contentsOfDirectory(at: docsDir, includingPropertiesForKeys: nil) {
            for url in items {
                let name = url.lastPathComponent
                if name == "hub_rooms.json" || name == "hub_graph.json" ||
                   name == "hub_unlinked_openings.json" || name.hasPrefix("hub_room_") {
                    try? fm.removeItem(at: url)
                }
            }
        }
    }
}
